### PR TITLE
Feature/fbfw 129 convoy

### PR DESCRIPTION
**Pull request by bryancodazen**
### Summary
This PR addresses scenarios 1&4 of the FBFW-129: https://codazen.atlassian.net/browse/FBFW-129
There are still kinks that need to be ironed out. As a part of the task I also refactored game api slice as the file was getting too large and difficult to read. 

These are the exact files which underwent changes or are new as they are relevant to what was needed for the Convoy move:

ArrowColor.ts
WDMapController.ts
ContextVar.ts
initial-state.ts
GameState.ts
GameStateMaps.ts
SaveOrders.ts
OrderType.ts
WDArrowMarkerDefs.tsx
drawConvoyOrders.ts
drawOrders.ts
getOrdersMeta.ts
generateMaps.ts
processConvoy.ts
webDiplomacyTheme.ts

Every other file is basically a function that was moved out of game-api-slice.ts.

**You need to update your .env files to the following: https://codazen.atlassian.net/wiki/spaces/FWD/pages/2549121027/Web+Diplomacy+Environment+information**

### Testing Plan

1. setup a game to have single and multi convoy paths. 
2. single click on an army
3. single click on a convoy chain
4. single click on the destination territory
5. A black arrow will be drawn from the army to the territory and blue lines will connect to the arrow from the fleets participating in the convoy. 

### Video
https://drive.google.com/file/d/1elv3y__cXNFW-EXqbVCv2mlwRb-AwClJ/view?usp=sharing

### CODE REVIEW

**Comment by mariomelchor** [comment]
Can we move the refactor to its own PR and then stack the convoy on top of that?
This is a Big PR and hard to follow what changes are actually happening for the convoy.

**Comment by bryancodazen** [comment]
> Can we move the refactor to its own PR and then stack the convoy on top of that? This is a Big PR and hard to follow what changes are actually happening for the convoy.

I noted that the changes for convoy are happening inside of `getOrdersMeta.ts` and `processConvoy.ts`. It's going to be a lot of work to split this PR up and the result will be the same amount of work having to be reviewed.

**Comment by mariomelchor** [comment]
> I noted that the changes for convoy are happening inside of `getOrdersMeta.ts` and `processConvoy.ts`. It's going to be a lot of work to split this PR up and the result will be the same amount of work having to be reviewed.

I agree it's the same amount of work but its a refactor and adding convoy support in 1 PR was asking purely for code review purposes since this is a Big PR.

**Comment by bryancodazen** [comment]
> 



> > I noted that the changes for convoy are happening inside of `getOrdersMeta.ts` and `processConvoy.ts`. It's going to be a lot of work to split this PR up and the result will be the same amount of work having to be reviewed.
> 
> I agree it's the same amount of work but its a refactor and adding convoy support in 1 PR was asking purely for code review purposes since this is a Big PR.

These are the exact files which underwent changes or are new as they are relevant to what was needed for the Convoy move: 
- ArrowColor.ts
- WDMapController.ts
- ContextVar.ts
- initial-state.ts
- GameState.ts
- GameStateMaps.ts
- SaveOrders.ts
- OrderType.ts
- WDArrowMarkerDefs.tsx
- drawConvoyOrders.ts
- drawOrders.ts
- getOrdersMeta.ts
- generateMaps.ts
- processConvoy.ts
- webDiplomacyTheme.ts

Every other file is basically a function that was moved out of game-api-slice.ts. As I was working through the convoy order logic I did this refactor simultaneously. It could've been its own task, but it helped in paving a way forward for convoy. At this point, splitting this up is more work than I think it's worth. I'd have to go back, cherry pick these files, put them in a new PR. Rebase the convoy on top of that one, and rebase the other two PRs that branch off this one. Not worth it IMO.

**Comment by justinyoocodazen** [comment]
image

Not sure if this is addressed in a later pr or intended behavior but the army in picardy can convoy move anywhere the north sea touches without clicking on the english channel.

image

Again, this might be out of the scope of this PR but when I then click the english channel and then Wales, the blue convoy arrow from the north sea follows, even though the north sea does not touch Wales. If this has been addressed in the later ones, sorry, but felt like I should type this finding out just in case.

**Comment by justinyoocodazen** [comment]
image

This is what was saved, which looks like it makes sense to me, so it'd just be removing the arrow which I'm guessing was implemented in the later prs.

When following the test plan everything looked good to me.

**Comment by bryancodazen** [comment]
> 

Yea, both of the mock issues are a result of the arrow still being connected to the previous order, without showing implied arrows. In the first mock it would show a gray arrow from the english channel fleet (since it is implied). The second mock would show an implied move arrow from the army in picardy to the previous order. (implied)

**Review by justinyoocodazen — APPROVED** [review]

File: beta-src/src/utils/state/gameApiSlice/reducers/processMapClick.ts [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+      subsequentClicks: GameState["order"]["subsequentClicks"];
+      toTerritory: GameState["order"]["toTerritory"];
+      type: GameState["order"]["type"];
+      unitID: GameState["order"]["unitID"];
+    };
+    ordersMeta: OrdersMeta;
+    overview: {
+      user: { member: GameOverviewResponse["user"]["member"] };
+      phase: GameOverviewResponse["phase"];
+    };
```
Comment (sariacodazen) [inline beta-src/src/utils/state/gameApiSlice/reducers/processMapClick.ts:62]: If you just typed data, order, and overview, then destructured their children after line 64, wouldnt the children be automatically typed without having to type all of them?

**Review by sariacodazen — APPROVED** [review]
The code looks good to me although I left a comment that may be cleaner. My main concern with this major refactor is how this can effect the code that is currently in, and about to be merged (since this is more clean up and those are more functional with deadlines). Do we plan on merging this after we have the bulk of needed functionality in the code base? Id hate to be on a wild refactor ride with such a tight deadline

**Comment by bryancodazen** [comment]
> The code looks good to me although I left a comment that may be cleaner. My main concern with this major refactor is how this can effect the code that is currently in, and about to be merged (since this is more clean up and those are more functional with deadlines). Do we plan on merging this after we have the bulk of needed functionality in the code base? Id hate to be on a wild refactor ride with such a tight deadline

I can resolve your merge conflicts if you feel it's going to hinder you
